.travis/find-unmatched-test-packages

```diff
@@ -0,0 +1,62 @@
+#!/bin/bash
+
+set -u
+
+red="\e[91m"
+bold="\e[1m"
+normal="\e[0m"
+
+status=0
+
+## Put all source and test files into a temporary file to serve as a cache
+testFiles=$(mktemp)
+srcFiles=$(mktemp)
+trap 'rm -rf $testFiles $srcFiles' EXIT
+
+find . -path "*src/test/java/*" -type f -name "*Test.java" > "$testFiles"
+find . -path "*src/main/java/*" -type f -name "*.java" > "$srcFiles"
+
+
+## Loop over all source files
+while read -r file; do
+     ## Find the expected test file for a given source file and count how
+     ## many times we find that test file name in the set of test files
+     testFileName=$(basename "${file/.java/Test.java}")
+     testFileMatchCount=$(grep -c "/$testFileName" "$testFiles")
+
+     ## If we find the expected test file name, check if test file package
+     ## matches the source file
+     if [ "$testFileMatchCount" == 1 ]; then
+
+       ## Grab the name of the matching test file with full path
+       testFileMatch=$(grep "/$testFileName" "$testFiles")
+
+       ## chop off the filename and the '/src/test/java/' prefix
+       ## EG: ./project/src/test/java/org/triplea/package/File.java -> org/triplea/package
+       testPackage=$(echo "$testFileMatch" | \
+            sed 's|\(.*\)/.*\.java$|\1|' | sed 's|.*/java/||')
+
+       ## chop off the filename and the '/src/main/java/' prefix
+       ## EG: ./project/src/main/java/org/triplea/package/File.java -> org/triplea/package
+       srcPackage=$(echo "$file" | \
+            sed 's|\(.*\)/.*\.java$|\1|' | sed 's|.*/java/||')
```

Comment:
This regex feels a little bit weird:
```regex
(.*)/.*\.java$
```

Because in theory for a string `/a/b/c.java`
Group 1 could be either one of
- `/a/b`
 - `/a`
 - ` `

Maybe there's a convention to regex engines, but I'd really feel a little bit safer if the
regex would look a little bit more like this:
```regex
(.*)/[^/]*\.java$
```
or
```regex
(.*)/.*?\.java$
```
Not sure what sed supports.
Same thing a couple of lines above

Reply:
I assume you're referring to the first regex @RoiEXLab .
`sed` does longest possible match. For `/a/b/c.java`, group 1 will be: `a/b`
I dont' quite follow the examples given, I feel like I should probably get more up to speed with javascript and/or perl regexes, I'm very used to GNU regex. Could you explain how those variants would be more robust?

Reply:
Yes,
```regex
[^/]
```

Would match all characters except the forward slash, so it is guaranteed to be just the filename without any of the directory stuff.
```regex
.*?
```
The `?` after the `*` is the lazy operator which explicitly makes this part try to match as few characters as possible, which will achieve the same result a little bit more explicit and not possibly rely on some undefined regex engine behaviour (maybe a different linux distro uses some other implementation of sed?)

Reply:
The first suggestion which guards against non-greedy matchers works.
```
dan@dan-desk:~/work/triplea$ echo "./test-common/src/main/java/org/triplea/test/common/security/TestSecurityUtils.java" | sed 's|\(.*\)/[^/]*\.java$|\1|'
./test-common/src/main/java/org/triplea/test/common/security
```

Unless I manged something, the second one does not:
```
dan@dan-desk:~/work/triplea$ echo "./test-common/src/main/java/org/triplea/test/common/security/TestSecurityUtils.java" | sed 's|\(.*\)/.*?\.java$|\1|'
./test-common/src/main/java/org/triplea/test/common/security/TestSecurityUtils.java
```

`sed` is greedy matching, that should not vary by distro. 

http://sed.sourceforge.net/sedfaq5.html

>  The RE '.*' is designed to be "greedy" (i.e., matching as many characters as possible)

5.4. My RE isn't matching/deleting what I want it to. (Or, "Greedy vs. stingy pattern matching")